### ISSUE TITLE
Automated browser tests docs and execution

### ISSUE THREAD

**Issue by th31nitiate** [nitrogen/nitrogen#145]
What have been working on is getting a bit out of hand.  It becoming a maintance burden if though is majorly still in development there so much going on that a small change in each place cause issues else were (It clearly obvious I'm working without a design spec or anything fancy just basic feature requests).

I'm starting to heavily consider implementing tests. The thing is eunit test's can be sufficient but browser tests are also an important part of it. There is the option of selinium which has web driver and a detailed guide [here](http://blog.equanimity.nl/blog/2014/03/07/selenium-webdriver-for-erlang/) . It a tiny bit of an overkill. I noticed nitrogen provides browser tests and I was wonder if you expand on how to set these up and I'll try update the documents when possible.

I noticed that  this `-include_lib("nitrogen_core/include/wf_test.hrl").` is not mentioned in the docs as a requirement to get test working as it contains the needed macros.

then when I included this and one test there seems to be issues giving the following errors when updating an element defined by both `id` and `html_name`: 

`
3> login:tests().
** exception error: an error occurred when evaluating an arithmetic expression
     in function  wf_test:increment_fail/0 (/home/m3rl1n/Code/app/_build/default/lib/nitrogen_core/src/lib/wf_test.erl, line 261)
     in call from wf_test:internal_fail/2 (/home/m3rl1n/Code/app/_build/default/lib/nitrogen_core/src/lib/wf_test.erl, line 250)
`  

My browser of choice will likely be Firefox but likely both. Trying to find a way to set up basic tests to run and then I can expand upon them when I start writing solid tests later on.

I think part of appreciating Erlang is learning why things are the way they are by experience the cause of their implementation. One good example of this is type specs and I feel many of my problem maybe solved by simple type specifications & obviously tests.

**Comment by choptastic** [comment]
Hey David!

I'm aware of Selenium and the Selenium Web Driver for erlang.  That said, because the erlang driver has a [modification](https://github.com/Quviq/webdrv/blob/master/LICENSE#L9) to its licence that disqualifies it from being a true open source license, I decided to implement my own testing system in Nitrogen. This ensures that 1) it adhere's the Nitrogen's open source MIT license, and 2) because it relies exclusively on native nitrogen functionality, no addons are required.

As for the wf_test.hrl file not being mentioned in the docs: [wf_test.hrl already is included in wf.hrl](https://github.com/nitrogen/nitrogen_core/blob/master/include/wf.hrl#L5), hence why the the documentation for the testing system doesn't mention it.

The error message you mentioned is interesting, 

For a few examples here that will help point you in the right direction, all of Nitrogen's actual tests can be found in the [NitrogenProject.com](https://github.com/nitrogen/NitrogenProject.com) repo.

Here are a few simpler examples: 

* A pretty simple set of tests for a basic self-contained page: https://github.com/nitrogen/NitrogenProject.com/blob/master/src/tests/tests_dropdown.erl
* Adding tests to a page that already exists without adding the code to that specific page: https://github.com/nitrogen/NitrogenProject.com/blob/master/src/tests/tests_advancedcontrols1.erl (this adds tests to https://github.com/nitrogen/NitrogenProject.com/blob/master/src/demos/demos_advancedcontrols1.erl through the use of 'wf_test:start_other(OtherModuleName, TestStartFun)`
* A somewhat more complicated test that integrates a bunch of javascript into the tests: https://github.com/nitrogen/NitrogenProject.com/blob/master/src/tests/tests_cookie.erl

Basically, Nitrogen's testing system is its own self-contained system that does not use eunit or ct, and launched by:

1) Adding some [new rules to the app.config](https://github.com/nitrogen/NitrogenProject.com/blob/master/app.config#L22-L64), then
2) Launching it from the erlang shell with `wf_test:start_all(nitrogen_core)` (you can see it launched from the NitrogenProject repo here: https://github.com/nitrogen/NitrogenProject.com/blob/rebar3/Makefile#L106-L107)

Hopefully that helps!

**Comment by th31nitiate** [comment]
Thank you. That is a great help. I will give this a go in a few hours and let you know if I manage to get it working running. 

I don't quite under this syntax "##textbox > something > something_else ", I just assume that's is selector syntax similar to xpath which in slightly more familiar with.

The selenium driver has a few hiccups such as the inability to manage cookies amongst a few other things. 

Though all in all this is great stuff.

**Comment by th31nitiate** [comment]
Having tried this it is logical for the most part and everything pretty much runs as intended. I'm happy that I'm able to at least to launch the browser. The issue is that its more directed at elements. This is find as it would be useful when I start developing the bulma plugin when I get some time. 

The issue with what I have now is simply that loading the test is not completely possible. Even when using the start other feature. For example is I  reference in `app.config`  `tests/web_test.erl`, it will literarily load that path. Though the page I maybe trying to test will be located at lets say `/login`.

From a plugin and element development side of things it makes sense. But the fully built application maybe a little bit more difficult. I would of liked to see the element get filled in by the value after the browser is launched.

All in all for now I'm going to have to focus solely on Eunit tests until I'm developing framework add-ons. 

I will also try see if I can add some of the missing tests detailed in the to-do part of some of those files. It will help me get better understand of whats going on in the code. 

I think maybe I'm trying to do to much for the most part. WebApp QA stuff should ideally be left to them or at least be dependent on their tool sets. I believe its called the UAT phase which is the complete far end of the pipeline to were we are sitting. The whole screen shot and visual regression stuff. It would be good if I did have a QA dept (or the do it your self mentality) but I don't so I'm just going to have to make do with what I do have.

Ok, you can close this for now if you want or leave open for others. I'm going to focus on unit tests for the moment with an emphasis on postbacks events. To try avoid errors I will use the validators to make sure I get the data I need when using `wf:q/1` or `wf:mq/1`. That should be good enough for now.

**Comment by choptastic** [comment]
Good to hear.
  
  If the `start_other` functionality doesn't work (as you mentioned in the login test, since it's looking for the `login` module), the other thing you could do to help would be to cheat by adding a module called something like `web_start_test.erl` that just sets a cookie or session variable like:
  
  ```erlang
  -module(web_start_test).
  -export([main/0]).
  main() ->
     wf:session(test_running, true),
     wf_test:start(fun() -> ok end),
     #template{file="your_template.html}.
  ```
  From there, you can throw a little check in each page's `main()` function:
  
  ```erlang
  main() ->
     ?WF_IF(wf:session(test_running)==true, wf_test:start(fun tests/0)),
     #template{file=YourTemplate}.
  ```
  
  That would give you some of the functionality you need here to have your tests explicitly run on your `login` page instead of a proxy page like `test_login`.
  
  > I would of liked to see the element get filled in by the value after the browser is launched.
  
  That part is handled by the `Setup` argument in the `test_X` functions.
  
  For example:
  
  ```erlang
Setup = fun() ->
   wf:set(login, "Test_username"),
   wf:set(password, "Test_password")
end,
Assertion = fun() ->
   [Name, PW] = wf:mq([login, password]),
   case db_login:attempt_login(Name, PW) of
      {ok, Loginid} ->
          true;
      _ ->
          false
    end
end,
?wf_test_auto(attempt_login, Setup, Assertion).
```

Anyway, hopefully that all makes sense, and that I'm answering the right questions.

I will go ahead and close this now, but if you have more questions, feel free to ask.
